### PR TITLE
Refactoring to be considered before adding MMTk

### PR DESCRIPTION
**Pull request by udesou**
This PR contains some refactoring of common functions that were moved to `gc-common.c` and should be shared between MMTk and Julia's stock GC.

### CODE REVIEW

File: src/staticdata.c
Comment (vchuravy) [inline src/staticdata.c:641]: This is new, best to separate from the rest of the refactoring.

**Comment by d-netto** [comment]
@udesou, a comment before the review: I think rebase might be preferable over merging master to sync a branch, since it avoids interleaving your commits with the upstream ones, and keeps the log "cleaner" (e.g. all of your commits after HEAD).

**Comment by udesou** [comment]
> @udesou, a comment before the review: I think rebase might be preferable over merging master to sync a branch, since it avoids interleaving your commits with the upstream ones, and keeps the log "cleaner" (e.g. all of your commits after HEAD).

Sounds good. I was just clicking the button on GitHub without changing anything but I think I did it right this time.

File: src/gc-stock.c
Comment (d-netto) [inline src/gc-stock.c:2801]: Should probably be kept in `gc-stock.h`, as discussed in a previous comment.

File: src/gc-stock.c
Comment (d-netto) [inline src/gc-stock.c:2790]: Should probably be kept in `gc-stock.h`, as discussed in a previous comment.

File: src/gc-stacks.c
Comment (d-netto) [inline src/gc-stacks.c:84]: Why no longer `static`? If MMTk is also using this function (which looks like the most reasonable guess), then we should consider adding a declaration to a header such as `julia_internal.h` (so you don't need to re-declare in MMTk itself).

File: src/gc-stacks.c
Comment (d-netto) [inline src/gc-stacks.c:49]: Same comment as the other implementation that's used in Unix systems.

File: src/gc-debug.c
Comment (d-netto) [inline src/gc-debug.c:1108]: Consider declaring it in `gc-common.h`? Since this header will be included by `gc-debug.c` either way...

File: src/gc-common.c
Comment (d-netto) [inline src/gc-common.c:689]: Putting everything here under `MISC` seems too coarse-grained.

Please consider creating additional categories in `gc-common.c` such as "allocation", "generic memory", "GC debug", etc for better code organization and readability.

**Comment by fingolfin** [comment]
@udesou if you don't mind, could you perhaps click "Resolve conversation" for comments that you already addressed? That would make it a tad easier to follow what's going on here. Thank you, and thank you for this work!

File: src/gc-common.c
Comment (d-netto) [inline src/gc-common.c:596]: Note that you are using the stock `gc_tls` in this function, even though this function is part of `gc-common.c` (i.e. it should work for any GC implementation).

Please consider refactoring it so that the abstraction is respected.

A possible approach would be to move `mafreelist` out of the `gc_tls` and into the `ptls` itself.

File: src/julia_internal.h
Comment (d-netto) [inline src/julia_internal.h:370]: NIT: no need for `extern` here?

File: src/gc-common.c
Comment (udesou) [reply src/gc-common.c:596]: What I have currently is having both `gc-tls` for stock and MMTk have a Julia `heap` (`jl_thread_heap_t`) field and both of them have a few common fields such as `weak_refs`, `live_tasks` and `mafreelist` for example. But for example, MMTk's one doesn't need the `remset` or `remset_nptr` fields. See [here](https://github.com/udesou/julia/blob/add-mmtk-upstream/src/gc-tls-mmtk.h).
Maybe a solution would be creating a common `gc_common_tls` and putting everything that is common within both GCs in there instead?

**Review by d-netto — CHANGES_REQUESTED** [review]
Please address the gcext failure on CI before merging.

**Comment by d-netto** [comment]
@fingolfin: I thought we moved gcext out of "Allow Fail" and into a regular test that needs to pass for CI to be green?

**Comment by udesou** [comment]
> Please address the gcext failure on CI before merging.

I think it should be fixed now, however, some unrelated things seem to be failing (by looking at CI results from the latest commit in the Julia repo). If it's really unrelated, then we should be good to go, I guess.

**Review by d-netto — APPROVED** [review]
Reading through it LGTM, but, as usual, let's wait for CI to merge...

**Comment by d-netto** [comment]
@udesou. CI looks fine. Could you fix the merge conflict in `src/gc-debug.c`?

**Pull request merged**
